test/test.js

```diff
@@ -0,0 +1,298 @@
+        return logger;
+      });
+      subject.register("logger", function(formatter) {
+        return formatter("message");
+      });
+      subject.register("formatter", function() {
+        return message => `formatted ${message}`;
+      });
+      assert.deepEqual(subject.get("app"), "formatted message");
+    });
+
+    it("should only initialize a dependency once", function() {
+      subject.register("logger", function(formatter) {
+        return { formatter };
+      });
+      subject.register("router", function(formatter) {
+        return { formatter };
+      });
+      subject.register("formatter", function() {
+        return message => `formatted ${message}`;
+      });
+      assert.equal(
+        subject.get("logger").formatter,
+        subject.get("router").formatter
+      );
+    });
+
+    it("should not clobber dependencies that are overridden", function() {
+      subject.register("logger", function(formatter) {
+        return formatter("message");
+      });
+      subject.register("formatter", function() {
+        return message => `formatted ${message}`;
+      });
+      assert.equal(
+        subject.get("logger", { formatter: () => "overridden message" }),
+        "overridden message"
+      );
+      assert.equal(subject.get("logger"), "formatted message");
+    });
+
+    context("errors", function() {
+      it("should throw error on circular dependency", function() {
+        subject.register("app", function(logger) {
+          return logger;
+        });
+        subject.register("logger", function(app) {
+          return app;
+        });
+        assert.throws(() => subject.get("app"), /circular dependency/);
+      });
+
+      it("should NOT throw circular dependency error if two modules require the same dependency", function() {
+        subject.register("app", function(logger, router) {
+          return [logger, router];
+        });
+        subject.register("logger", function(formatter) {
+          return formatter("message");
+        });
+        subject.register("router", function(formatter) {
+          return formatter("route");
+        });
+        subject.register("formatter", function() {
+          return message => `formatted ${message}`;
+        });
+        assert.deepEqual(subject.get("app"), [
+          "formatted message",
+          "formatted route"
+        ]);
+      });
+
+      it("should throw an error if it cant find a dependency", function() {
+        assert.throws(() => subject.get("goats"));
+      });
+
+      it("should throw an error if it cant find a transitive dependency", function() {
+        subject.register("goats", function(cans) {
+          return cans;
+        });
+        assert.throws(() => subject.get("goats"));
+      });
+    });
+
+    context("internals", function() {
+      it('should register the container as "_container"', function() {
+        assert.equal(subject.get("_container"), subject);
```

Comment:
What is the use case for this feature?

Reply:
Not sure, I believe the original intent was to expose internals if you wanted to do some sort of meta programming around the guts of the container?

Reply:
I vote for dropping this until a clear use case presents itself. Even then, if metaprogramming is that desirable, I'd rather an explicit "plugin" API.